azure/services/virtualmachines/virtualmachines.go

```diff
@@ -82,211 +76,57 @@ func (s *Service) Reconcile(ctx context.Context) error {
 	ctx, span := tele.Tracer().Start(ctx, "virtualmachines.Service.Reconcile")
 	defer span.End()
 
-	vmSpec := s.Scope.VMSpec()
-	existingVM, err := s.getExisting(ctx, vmSpec.Name)
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureServiceReconcileTimeout)
+	defer cancel()
 
-	switch {
-	// VM got deleted outside of capz
-	case err != nil && azure.ResourceNotFound(err) && s.Scope.ProviderID() != "":
-		s.Scope.SetVMState(infrav1.Deleted)
-		return azure.VMDeletedError{ProviderID: s.Scope.ProviderID()}
-	case err != nil && !azure.ResourceNotFound(err):
-		return errors.Wrapf(err, "failed to get VM %s", vmSpec.Name)
-	case err == nil:
-		// VM already exists, update the spec and skip creation.
-		s.Scope.SetProviderID(azure.ProviderIDPrefix + existingVM.ID)
-		s.Scope.SetAnnotation("cluster-api-provider-azure", "true")
-		s.Scope.SetAddresses(existingVM.Addresses)
-		s.Scope.SetVMState(existingVM.State)
-		s.Scope.UpdateStatus()
-	default:
-		s.Scope.V(2).Info("creating VM", "vm", vmSpec.Name)
-		sku, err := s.resourceSKUCache.Get(ctx, vmSpec.Size, resourceskus.VirtualMachines)
-		if err != nil {
-			return azure.WithTerminalError(errors.Wrapf(err, "failed to get SKU %s in compute api", vmSpec.Size))
-		}
+	vmSpec := s.Scope.VMSpec()
 
-		storageProfile, err := s.generateStorageProfile(ctx, vmSpec, sku)
-		if err != nil {
-			return err
+	result, err := async.CreateResource(ctx, s.Scope, s.Client, vmSpec, serviceName)
+	s.Scope.UpdatePutStatus(infrav1.VMRunningCondition, serviceName, err)
+	if err == nil && result != nil {
+		vm, ok := result.(compute.VirtualMachine)
+		if !ok {
+			return errors.Errorf("%T is not a compute.VirtualMachine", result)
 		}
-
-		securityProfile, err := getSecurityProfile(vmSpec, sku)
+		infraVM, err := converters.SDKToVM(vm)
 		if err != nil {
 			return err
 		}
+		s.Scope.SetProviderID(azure.ProviderIDPrefix + infraVM.ID)
+		s.Scope.SetAnnotation("cluster-api-provider-azure", "true")
 
-		nicRefs := make([]compute.NetworkInterfaceReference, len(vmSpec.NICNames))
-		for i, nicName := range vmSpec.NICNames {
-			primary := i == 0
-			nicRefs[i] = compute.NetworkInterfaceReference{
-				ID: to.StringPtr(azure.NetworkInterfaceID(s.Scope.SubscriptionID(), s.Scope.ResourceGroup(), nicName)),
-				NetworkInterfaceReferenceProperties: &compute.NetworkInterfaceReferenceProperties{
-					Primary: to.BoolPtr(primary),
-				},
-			}
-		}
-
-		priority, evictionPolicy, billingProfile, err := converters.GetSpotVMOptions(vmSpec.SpotVMOptions)
-		if err != nil {
-			return errors.Wrap(err, "failed to get Spot VM options")
-		}
-
-		osProfile, err := s.generateOSProfile(ctx, vmSpec)
+		// Discover addresses for NICs associated with the VM
+		addresses, err := s.getAddresses(ctx, vm, vmSpec.ResourceGroupName())
```

Comment:
future work I'm considering: figure out how to do this without needing to get the NICs and IPs again (we created them earlier), but this would be another refactor so holding off for now since this PR already does a lot. Kept that function as-is.